packages/contracts/evm-contracts/contracts/orderbook/OrderbookDex.sol

```diff
@@ -91,27 +123,51 @@ contract OrderbookDex is IOrderbookDex, ERC1155Holder, ReentrancyGuard {
 
     /// @inheritdoc IOrderbookDex
     function fillOrdersExactEth(
+        address asset,
         uint256 minimumAsset,
         uint256[] memory orderIds
-    ) public payable virtual nonReentrant {
+    ) external payable virtual nonReentrant {
         uint256 length = orderIds.length;
         uint256 remainingEth = msg.value;
         uint256 totalAssetReceived;
         for (uint256 i; i < length; ++i) {
             uint256 orderId = orderIds.unsafeMemoryAccess(i);
-            Order storage order = orders[orderId];
+            Order storage order = orders[asset][orderId];
             if (order.assetAmount == 0) {
                 continue;
             }
-            uint256 assetsToBuy = remainingEth / order.pricePerAsset;
+
+            uint256 purchaseCost = (remainingEth * basisPoints) / (order.takerFee + basisPoints);
+            if (purchaseCost * (order.takerFee + basisPoints) != (remainingEth * basisPoints)) {
```

Comment:
we should probably add a comment explaining what this line is for

Reply:
Added